executor/infoschema_reader.go

```diff
@@ -2350,6 +2324,143 @@ func (r *dataLockWaitsTableRetriever) retrieve(ctx context.Context, sctx session
 	return res, nil
 }
 
+type deadlocksTableRetriever struct {
+	dummyCloser
+	batchRetrieverHelper
+
+	currentIdx          int
+	currentWaitChainIdx int
+
+	table       *model.TableInfo
+	columns     []*model.ColumnInfo
+	deadlocks   []*deadlockhistory.DeadlockRecord
+	initialized bool
+}
+
+func (r *deadlocksTableRetriever) nextIndexPair(idx, waitChainIdx int) (int, int) {
+	waitChainIdx++
+	if waitChainIdx >= len(r.deadlocks[idx].WaitChain) {
+		waitChainIdx = 0
+		idx++
+		for idx < len(r.deadlocks) && len(r.deadlocks[idx].WaitChain) == 0 {
+			idx++
+		}
+	}
+	return idx, waitChainIdx
+}
+
+func (r *deadlocksTableRetriever) retrieve(ctx context.Context, sctx sessionctx.Context) ([][]types.Datum, error) {
```

Comment:
Is it possible to abstract the retrieve implementation together with the one used by `TiDB_TRX` table?

Reply:
It's kind of hard... Can we just continue this PR, and I submit another PR if I find some way to abstract it?